dali/test/python/test_external_source_parallel.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2020-2021, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from nose.tools import raises, with_setup
+
+from test_pool_utils import *
+from test_utils import compare_pipelines
+from test_external_source_parallel_utils import *
+
+
+# Test that we can launch several CPU-only pipelines by fork as we don't touch CUDA context.
+@with_setup(setup_function, teardown_function)
+def test_parallel_fork_cpu_only():
+    global pipe_processes
+    pipeline_pairs = 4
+    batch_size = 10
+    iters = 40
+    callback = ExtCallback((4, 5), iters * batch_size, np.int32)
+    parallel_pipes = [(create_pipe(callback, 'cpu', batch_size, py_num_workers=4,
+                                   py_start_method='fork', parallel=True, device_id=None),
+                       create_pipe(callback, 'cpu', batch_size, py_num_workers=4,
+                                   py_start_method='fork', parallel=True, device_id=None))
+                      for i in range(pipeline_pairs)]
+    for pipe0, pipe1 in parallel_pipes:
+        pipe0.build()
+        pipe1.build()
+        capture_processes(pipe0._py_pool)
+        capture_processes(pipe1._py_pool)
+        compare_pipelines(pipe0, pipe1, batch_size, iters)
+
+@with_setup(setup_function, teardown_function)
+def test_parallel_no_workers():
+    batch_size = 10
+    iters = 4
+    callback = ExtCallback((4, 5), iters * batch_size, np.int32)
+    parallel_pipe = create_pipe(callback, 'cpu', batch_size, py_num_workers=0,
+                                py_start_method='spawn', parallel=True, device_id=None)
+    parallel_pipe.build()
+    capture_processes(parallel_pipe._py_pool)
+    assert parallel_pipe._py_pool is None
+    assert parallel_pipe._py_pool_started == False
+
+@with_setup(setup_function, teardown_function)
+def test_parallel_fork():
+    epoch_size = 250
+    callback = ExtCallback((4, 5), epoch_size, np.int32)
+    pipes = [(
+        create_pipe(
+            callback, 'cpu', batch_size, py_num_workers=num_workers, py_start_method='fork',
+            parallel=True),
+        create_pipe(callback, 'cpu', batch_size, parallel=False),
+        dtype, batch_size)
+        for dtype in [np.float32, np.int16]
+        for num_workers in [1, 3, 4] for batch_size in [1, 16, 150, 250]]
```

Comment:
Won't you have a lot of pipelines created here at once?
Maybe you should create one after another?

Reply:
We need to start all of them before we run them, otherwise we touched CUDA and we hit the assert.